capsules/src/process_console.rs

```diff
@@ -224,6 +243,61 @@ pub struct ProcessConsole<'a, A: Alarm<'a>, C: ProcessManagementCapability> {
     capability: C,
 }
 
+#[derive(Copy, Clone)]
+pub struct Command {
+    buf: [u8; COMMAND_BUF_LEN],
+    len: usize,
+}
+
+impl Command {
+    pub fn new() -> Command {
+        Command {
+            buf: [0; COMMAND_BUF_LEN],
+            len: 0,
+        }
+    }
+
+    /// Fill the buffer with the provided data.
+    /// If the provided data's length is smaller than the buffer length,
+    /// the left over bytes are not modified due to '\0' termination.
+    pub fn fill(&mut self, buf: &[u8; COMMAND_BUF_LEN], terminator_idx: usize) {
+        self.len = if terminator_idx >= COMMAND_BUF_LEN {
+            COMMAND_BUF_LEN
+        } else {
+            terminator_idx
+        };
+
+        (&mut self.buf).copy_from_slice(buf);
+    }
+
+    pub fn is_buffer_empty(&mut self) -> bool {
+        self.len == 0
+    }
+
+    pub fn get_byte(&self, byte_idx: usize) -> Option<u8> {
+        if byte_idx >= self.len {
+            return None;
+        }
+
+        Some(self.buf[byte_idx])
+    }
+}
+
+impl Default for Command {
+    fn default() -> Self {
+        Command {
+            buf: [0; COMMAND_BUF_LEN],
+            len: 0,
+        }
+    }
+}
+
+impl PartialEq<[u8; COMMAND_BUF_LEN]> for Command {
+    fn eq(&self, other_buf: &[u8; COMMAND_BUF_LEN]) -> bool {
+        self.buf.iter().zip(other_buf.iter()).all(|(a, b)| a == b)
```

Comment:
```suggestion
        self.buf.iter().zip(other_buf.iter()).all(|(a, b)| a != '\0' && b != '\0' && a == b)
```

Reply:
Done

Reply:
> It should stop after the first false https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.all. Write it in a comment.

If we add the *a != '\0' && *b != '\0' will return always false

Reply:
I found a method using the take_while https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/core/iter/trait.Iterator.html#method.take_while
In order to take just the elements until hit the NULL Terminator and then to compare them, however I do not know if it bring a big overload when comparing, basically it does two comparisions one to take the element and then one to compare the elements, which doubles the number of comparisions

Reply:
Or we could do something like this
```RUST
self.buf
            .iter()
            .zip(other_buf.iter())
            .take_while(|(a, b)| a == b && **a != TERMINATOR)
            .count()
            == self.len - 1
```
however this also doe snot work

Reply:
other_buf might be shorter than self.buf.

Reply:
> If we add the *a != '\0' && *b != '\0' will return always false

Why?

Reply:
From the documentation the all stops at the first false and returns false and if predicate returned true every time returns true, however from the condition `*a != '\0' && *b != '\0' && *a == *b`, when the all will encounter a NULL Termiantor it will retun false and not true even though all of the last predicates returned true

Reply:
> other_buf might be shorter than self.buf.

From the function signature the other_buf should match and array of COMMAND_BUF_LEN

Reply:
Perhaps this code could solve the issue, I am not sure about this:
```RUST
let mut try_iter = self.buf.iter().zip(other_buf.iter());

try_iter.all(|(a, b)| *a == *b && *a != TERMINATOR);

try_iter.next().eq(&Some((&TERMINATOR, &TERMINATOR)))
```
First we check if all of the elements are equal until hit a NULL terminator, the try_iter.next() can not be None because we stoped until first false so the next item should be either a pair of (a, b) or a pair of null terminators, and then we make sure that the next pair indeed is a pair of NULL Terminators, what do you think about this?

Reply:
Tested on the board and worked